[LoRaWAN]: Adding QOS in response to LinkADRReq and fixing class C bugs

hasnainvirk opened the pull request:
### Description

<!-- 
    Required
    Add here detailed changes summary, testing results, dependencies 
    Good example: https://os.mbed.com/docs/latest/reference/workflow.html (Pull request template)
-->
Adds QOS mechanism for unconfirmed messages as per 1.0.2 spec in response to LinkADRReq mac command. In addition to that it fixes various bugs potentially hindering class C operation.
Unit tests for missing methods and fixes for broken logic in the unit tests is also added.

### Pull request type

<!-- 
    Required
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front
    of them would change the meaning incorrectly. The only changes to be made are to add a description text under the
    description heading and to add a 'x' to the correct box.
-->
    [X] Fix
    [ ] Refactor
    [ ] Target update
    [ ] Functionality change
    [ ] Docs update
    [X] Test update
    [ ] Breaking change

Commit:
Style correction

LinkADRReq parameters and certain parameters used camel case
which is not the recommended style.

Commit:
Adding check for ongoing automatic-uplink

Before going after an automatic uplink, we should check if there was an
automatic uplink already ongoing, i.e., the ack for the previous
automatic uplink cycle has not been received.
If there is we shouldn't queue the new automatic uplink and wait for
the previous Ack cycle to complete.

Commit:
Adding QOS handling and fixing bugs for Class C

LinkADRReq mac command can be used by the network server to set a
certain level of QOS using NbTrans field which is applicable to
Unconfirmed traffic only for 1.0.2 spec.
This commit introduces mechanisms to facilitate this QOS. It means to
repeat an outgoing unconfirmed message NbTrans times without changing
its frame counter.

For class C, we have retired the ack_expiry_timer_for_class_c and have
replaced it with another timer which mimics the RX2 closure as in Class
A but doesn't actually close RX2 window. It's just a mechanism by which
the state machine is informed that the you can proceed forward, we have
not received anything in RX2 window either. This is needed as RX2
doesn't timeout in class C (i.e., the radio remains in continuous mode).
In addition to that we need to close any pending timers for Receive
windows after the MIC has passed and the Duplicate counter check has
also been passed.

Commit:
Making sure that RX slots open after state change

After transmission we should change the state before invoking opening of
slots as we may start receiving in the rx slots and the state would
suddenly change from SENDING to RECEIVING without going through the
ACK_WAIT state (in case of CONFIRMED messages). Tests show that after
this slight adjustment, our number of ack retries have significantly
reduced.

Commit:
TX post-process for CONFIRMED UL in no-reception case

The idea behind the method post_process_no_reception() is to post
process any outgoing TX but we shouldn't do that if a CONFIRMED message
is outgoing and there are still some retries left.

Commit:
Proper handling of RX1 frequency in rx_config

Previously, we weren't filling in RX1 frequecny in rx_window1_config
structure. However, everything worked as in LoRaPHY::rx_config() API
there was a check which filled in correct RX1 frequency.
Now we are filling in RX1 freq. properly while we are computing
parameters for RX1 window.

Commit:
Fixing coverity findings

A couple of the coverity analysis findings are being treated here. For
the rest there will be a separate PR.

Commit:
Unit Test Fixes for LoRaWAN

Missing methods are added.
Logic was broken at various places which needed to be fixed.

hasnainvirk commented:
@AnttiKauppila, @cmonr  Please review.

cmonr commented:
@hasnainvirk Please  take a look at the unittest. Looks like one of them is producing a segfault.

Commit:
Unit test fix for cancel_send()

PR 8299 went in but correct unit tests were not there.
This commit is adding proper unittests for 8299.

hasnainvirk commented:
@cmonr This commit will fix master https://github.com/ARMmbed/mbed-os/pull/8405/commits/8f0df13f0b940e092d67265649789b341fbe8134. Please don't revert the PR #8299 . Merge this one instead.

NirSonnenschein commented:
@hasnainvirk I re-ran the unit-test check and that still failed, then tried re-build on Jenkins as @OPpuolitaival  suggested , but it still failed: 
rerun: https://mbed-os.mbedcloudtesting.com/job/mbed-os-ci_unittests/366/
re-build: https://mbed-os.mbedcloudtesting.com/job/mbed-os-ci_unittests/367/
am I missing something?

0xc0170 commented:
> @cmonr This commit will fix master 8f0df13. Please don't revert the PR #8299 . Merge this one instead.

Still failing, can you review the failures?

@AnttiKauppila please review

cmonr commented:
/morph build

cmonr reviewed (approved):
Really hoping the rest of the PR does't fail the rest of CI.
